### ISSUE TITLE
IN pq.Array fails

### ISSUE THREAD

**Issue by psankar** [lib/pq#600]
I have the following code. Both the statements should return the same output, but I get `B,C` in the first and `A,B,C` in the second. The IN command fails to substitute with the pq.Array fields.

```
package main

import (
    "database/sql"
    "fmt"
    "github.com/lib/pq"
)

const (
    DB_USER     = "<username>"
    DB_PASSWORD = "<password>"
    DB_NAME     = "<db>"
)

func main() {
    dbinfo := fmt.Sprintf("user=%s password=%s dbname=%s sslmode=disable",
    DB_USER, DB_PASSWORD, DB_NAME)

    db, err := sql.Open("postgres", dbinfo)
    checkErr(err)
    defer db.Close()

    fmt.Println("# Querying with blacklist in SQL")
    rows, err := db.Query(`
    SELECT * FROM(
        SELECT x from (values ('A'), ('B'), ('C') ) s(x)
    ) As Res1 WHERE x NOT IN ('A');
`)
    checkErr(err)

    for rows.Next() {
        var str string
        err = rows.Scan(&str)
        fmt.Println(str) // Prints B,C
    }

    fmt.Println("Querying with blacklist in Golang")
    blacklist := []string{"A"}

    q := `
    SELECT * FROM(
        SELECT x from (values ('A'), ('B'), ('C') ) s(x)
    ) As Res1 WHERE x NOT IN ($1);

`
    rows, err = db.Query(q, pq.Array(blacklist))
    checkErr(err)

    for rows.Next() {
        var str string
        err = rows.Scan(&str)
        fmt.Println(str) // Prints A, B, C
    }
}

func checkErr(err error) {
    if err != nil {
        panic(err)
    }
}
```

The output is:
```
# Querying with blacklist in SQL
B
C
Querying with blacklist in Golang
A
B
C
```

**Comment by cbandy** [comment]
tl;dr `WHERE NOT x = ANY($1)` or `WHERE x <> ALL($1)`

----

1. PostgreSQL decides that `values ('A'), ('B'), ('C')` is `text`:
    ```
    => SELECT pg_typeof(x) FROM ( values ('A'), ('B'), ('C') ) s(x);
     pg_typeof 
    -----------
     text
     text
     text
    (3 rows)
    ```

2. The blacklist arrives at PostgreSQL as untyped literal `{A,B}` which is also happily treated as text; that is, one text value containing brackets and commas.

When PostgreSQL is misinterpreting parameters, it is a good idea to add typecast. In this case, changing to `$1::text[]` reveals the error: `pq: operator does not exist: text <> text[]`

To compare single values against the contents of an array, use the [`ANY()` and `ALL()` functions](https://www.postgresql.org/docs/current/static/functions-comparisons.html).

(See also #515, where integer types revealed the error much sooner.)

**Comment by psankar** [comment]
Thank you so much. I was able to get this working, with the `x <> ALL ($1)` construct. Shall I close the issue ?

**Comment by cbandy** [comment]
Yes, please!
